File: core/src/main/java/com/triangleleft/flashcards/mvp/main/MainPresenter.java
package com.triangleleft.flashcards.mvp.main;

import android.support.annotation.NonNull;

import com.annimon.stream.Collectors;
import com.annimon.stream.Optional;
import com.annimon.stream.Stream;
import com.triangleleft.flashcards.mvp.common.di.scope.ActivityScope;
import com.triangleleft.flashcards.mvp.common.presenter.AbstractPresenter;
import com.triangleleft.flashcards.mvp.vocabular.VocabularyNavigator;
import com.triangleleft.flashcards.service.account.AccountModule;
import com.triangleleft.flashcards.service.settings.Language;
import com.triangleleft.flashcards.service.settings.SettingsModule;
import com.triangleleft.flashcards.service.settings.UserData;
import com.triangleleft.flashcards.service.vocabular.VocabularyWord;
import com.triangleleft.flashcards.util.FunctionsAreNonnullByDefault;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Comparator;
import java.util.List;

import javax.inject.Inject;

import rx.Scheduler;

@FunctionsAreNonnullByDefault
@ActivityScope
public class MainPresenter extends AbstractPresenter<IMainView> implements VocabularyNavigator {

    private static final Logger logger = LoggerFactory.getLogger(MainPresenter.class);
    private final AccountModule accountModule;
    private final SettingsModule settingsModule;
    private final Scheduler scheduler;
    private final Comparator<Language> languageComparator =
            (l1, l2) -> Boolean.valueOf(l2.isCurrentLearning()).compareTo(l1.isCurrentLearning());

    private Language currentLanguage;
    private IMainView.Page currentPage;

    @Inject
    public MainPresenter(AccountModule accountModule, SettingsModule settingsModule, Scheduler scheduler) {
        super(IMainView.class);
        this.accountModule = accountModule;
        this.settingsModule = settingsModule;
        this.scheduler = scheduler;
    }

    @Override
    public void onCreate() {
        // Update user data here?
        applyState(view -> {
            currentPage = IMainView.Page.LIST;
            view.showList();
        });
    }

    @Override
    public void onBind(IMainView view) {
        super.onBind(view);
        Optional<UserData> userData = accountModule.getUserData();
        // It's possible that local data was wiped
        if (userData.isPresent()) {
            showUserData(userData.get());
        } else {
            getView().navigateToLogin();
        }
    }

    @Override
    public void showWord(@NonNull Optional<VocabularyWord> word) {
        logger.debug("showWord() called with: word = [{}]", word);
        applyState(view -> {
            currentPage = IMainView.Page.WORD;
            view.showWord(word);
        });
    }

    public void onBackPressed() {
        if (currentPage == IMainView.Page.LIST) {
            // list is considered top-level screen, so pressing back here should close app
            getView().finish();
        } else {
            // Go to top-level screen
            applyState(view -> {
                currentPage = IMainView.Page.LIST;
                view.showList();
            });
        }
    }

    public void onLanguageSelected(Language language) {
        if (language.equals(currentLanguage)) {
            return;
        }
        applyState(IMainView::showDrawerProgress);
        settingsModule.switchLanguage(language)
                .switchMap(nothing -> settingsModule.loadUserData())
                .observeOn(scheduler)
                .subscribe(data -> {
                            showUserData(data);
                            getView().reloadList();
                        },
                        error -> {
                            logger.error("switchLanguage()", error);
                            applyState(IMainView::showDrawerError);
                        }
                );
    }

    private void showUserData(UserData userData) {
        // We assume that is only one language that we are currently learning
        // Sort it, so it is always top of the list
        List<Language> languages = Stream.of(userData.getLanguages())
                .filter(Language::isLearning)
                .sorted(languageComparator)
                .collect(Collectors.toList());

        currentLanguage = languages.size() > 0 ? languages.get(0) : null;

        getView().showUserData(userData.getUsername(), userData.getAvatar(), languages);
    }

    public void onLogoutClick() {
        accountModule.setRememberUser(false);
        getView().navigateToLogin();
    }
}

File: app/src/main/kotlin/com/triangleleft/flashcards/ui/main/MainPresenter.kt
package com.triangleleft.flashcards.ui.main

import com.triangleleft.flashcards.di.scope.ActivityScope
import com.triangleleft.flashcards.service.account.AccountModule
import com.triangleleft.flashcards.service.settings.Language
import com.triangleleft.flashcards.service.vocabular.VocabularyWord
import com.triangleleft.flashcards.ui.ViewEvent
import com.triangleleft.flashcards.ui.common.presenter.AbstractRxPresenter
import com.triangleleft.flashcards.ui.vocabular.word.VocabularyNavigator
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.subjects.PublishSubject
import org.slf4j.LoggerFactory
import javax.inject.Inject
import javax.inject.Named

@ActivityScope
class MainPresenter @Inject constructor(
        accountModule: AccountModule,
        @Named(AbstractRxPresenter.UI_SCHEDULER) scheduler: Scheduler)
    : AbstractRxPresenter<IMainView, MainViewState, ViewEvent>(scheduler), VocabularyNavigator {


    private val wordClicks = PublishSubject.create<VocabularyWord>()
    private val backPresses = PublishSubject.create<Any>()
    private val initialState: MainViewState

    init {
        val userData = accountModule.userData.get()

        // It's possible that we are using account that doesn't have any languages
        val title = userData.currentLearningLanguage?.name ?: ""

        initialState = MainViewState(title, false, MainViewState.Page.List)
    }

    override fun onCreate(savedViewState: MainViewState?) {
        val initialState = savedViewState ?: initialState

        Observable.merge(
                backPresses.map { MainViewActions.backress() },
                wordClicks.map { MainViewActions.wordClick(it) })
                .scan(initialState) { state, action -> action.reduce(state) }
                .distinctUntilChanged()
                .subscribe { viewStates.onNext(it) }
    }

    override fun showWord(word: VocabularyWord) {
        wordClicks.onNext(word)
    }

    fun onLanguageChanged(currentLanguage: Language) {
        logger.debug("onLanguageChanged() called with: currentLanguage = [{}]", currentLanguage)
//        title = currentLanguage.name
//        view.setTitle(title)
//        view.reloadList()
    }


    private enum class Page {
        LIST,
        WORD
    }

    companion object {

        private val logger = LoggerFactory.getLogger(MainPresenter::class.java)
    }
}
